lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableMetricContainer.cs

```diff
@@ -0,0 +1,118 @@
+{
+    /// <summary>
+    /// A container containing collection of particular kind of mutable metrics
+    /// like counters, gauge etc.
+    /// </summary>
+    /// <typeparam name="T">Type of mutable metric.</typeparam>
+    public sealed class MutableMetricContainer<T> where T : MutableMetricBase
+    {
+        private readonly Dictionary<string, T> _metricsDictionary = new Dictionary<string, T>();
+        private readonly Func<string, T> _createMetricFromName;
+        private readonly Func<string, string, T> _createMetricFromNameDesc;
+        private readonly IObservable<SnapshotRequest> _provider;
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="metricFromName">A function that creates mutable metric from name.</param>
+        /// <param name="metricFromNameDesc">A function that creates mutable metric from nme and desc.</param>
+        /// <param name="provider">Observale to which new metric should subscribe to.</param>
+        public MutableMetricContainer(Func<string, T> metricFromName,
+            Func<string, string, T> metricFromNameDesc,
+            IObservable<SnapshotRequest> provider)
+        {
+            _createMetricFromName = metricFromName;
+            _createMetricFromNameDesc = metricFromNameDesc;
+            _provider = provider;
+        }
+
+        /// <summary>
+        /// String based indexer. Creates, gets and sets metric by name.
+        /// </summary>
+        /// <param name="name">Name of the metric,</param>
+        /// <returns>Existing or new metric.</returns>
+        public T this[string name]
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    if (!_metricsDictionary.ContainsKey(name))
+                    {
+                        _metricsDictionary[name] = _createMetricFromName(name);
+                        _metricsDictionary[name].Subscribe(_provider);
+                    }
+                    return _metricsDictionary[name];
+                }
+            }
+            set
+            {
+                lock (_lock)
+                {
+                    if (!_metricsDictionary.ContainsKey(name))
+                    {
+                        throw new MetricsException(string.Format("Metric with name {0} does not exist", name));
+                    }
+                    _metricsDictionary[name] = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// String based double indexer. Creates metircs vy name and description.
+        /// Gets and sets existing metric only by name.
+        /// </summary>
+        /// <param name="name">Name of the metric,</param>
+        /// <param name="desc">Description of the metric,</param>
+        /// <returns>Existing or new metric.</returns>
+        public T this[string name, string desc]
```

Comment:
So if metric does not exit, in the beginning when we want to create it for first time we might also want to give description of the metric apart from name. That why I decided to have double indexing. However, name is the unique way by which we identify the metric. Personally, I am not satisfied with it. An alternative might be to ask users to always explicitly create and add metric to the source but subsequent calls can be via index. Another alternative is to assume that via indexing descriptions cannot be provided and we can support additional API like AddCounter(name, descr) etc. 

Any suggestions?